### DISCUSSION TITLE
Issue with Subscribing to Supabase Channel for Authenticated Users

### DISCUSSION THREAD

**Discussion by shige0309** [supabase/supabase#29013] [Questions] [answered]
Hello,

I am currently using Supabase to ensure that only authenticated users can access data. However, I am facing an issue where authenticated users are unable to subscribe to a channel.

Here is the code I am using:

src/utils/supabase/server.ts

import { createServerClient, type CookieOptions } from "@supabase/ssr";
import { cookies } from "next/headers";
export function createClient(accessToken?: string) {
const cookieStore = cookies();
return createServerClient(
process.env.NEXT_PUBLIC_SUPABASE_URL!,
process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!,
{
global: {
headers: accessToken ? { Authorization: Bearer ${accessToken} } : {},
},
cookies: {
getAll() {
return cookieStore.getAll();
},
setAll(
cookiesToSet: {
name: string;
value: string;
options?: CookieOptions;
}[]
) {
try {
cookiesToSet.forEach(({ name, value, options }) =>
cookieStore.set(name, value, options)
);
} catch (error) {
console.error("Error setting cookies:", error);
}
},
},
}
);
}

src/app/api/subscribeLinkUsers/route.ts
const subscribeToChannel = async () => {
const channel = supabase
.channel("linkUsersChannel")
.on(
"postgres_changes",
{
event: "",
schema: "public",
table: "link_users",
},
async (payload) => {
console.log("Change received!", payload);
}
)
.subscribe();
setTimeout(() => {
channel.unsubscribe();
subscribeToChannel();
}, 60000);
};

Additionally, I have set the following RLS policy:

ALTER POLICY "Enable select for authenticated users only"
ON "public"."link_users"
TO authenticated
USING (
true
);

Issue Details:
- Authenticated users cannot subscribe to the channel
- No error messages are displayed, but the subscription does not work

Solutions Tried:
- Rechecked RLS policy
- Verified Supabase client settings
- Added error handling

I would appreciate any assistance you can provide.

Thank you.

**Comment by GaryAustin1** [comment, marked as answer]
Does the realtime monitor in the dashboard work for an authenticated user?

If so you likely don't have a user session in the client being used to setup/do realtime.
I don't know next.js or ssr that well, but can you do realtime in a server client?  Also you seem to be setting the Authorization header directly.  I thought SSR used cookies and handled setting the session for you.  If you set the Authorization header directly  that will not get to realtime as it uses the internal session info in the client, which setting the header bypasses.

**Reply by shige0309** [reply]
Thank you for your response.

It seems that the issue was with the Auth authentication part.
After adding supabase.realtime.setAuth(accessToken);, it worked fine.

Sorry for the trouble.
